### PR TITLE
Series performance improvements

### PR DESCRIPTION
**Pull request by stevezau**
### Motivation for changes:
If 100's of series exist flexget's performance can suffer when parsing. 

### Detailed changes:
- [x] Prefetch Series Data
- [x] Better Logic to parsing, right now it loops every entry * configured series..

### CODE REVIEW

**Comment by stevezau** [comment]
@gazpachoking @liiight cau you take a look at this? not sure other ways we can improve the speeds?

**Comment by gazpachoking** [comment]
Hmm. Is all the manual normalization needed? `Series.name` is meant to automatically compare with normalized version of the name. How much of a speed difference does this make?

**Comment by stevezau** [comment]
@gazpachoking it didn't work with name.in_. Do you know how to handle that in https://github.com/Flexget/Flexget/blob/develop/flexget/plugins/filter/series.py#L236-L238

**Comment by stevezau** [comment]
With pre-fetch it cut down the process time by 50% for me.

It's still very slow though. if you have an rss feed that has 100 entires and series with 100-200 entries it loops over the 100 entries times by the amount of series you have.. Can you think of a better way? I couldn't.

**Comment by gazpachoking** [comment]
Perhaps we need to implement this on the comparator? http://docs.sqlalchemy.org/en/latest/orm/internals.html#sqlalchemy.orm.properties.ColumnProperty.Comparator.in_

**Comment by stevezau** [comment]
ok, i'll look into that.

Any other idea how we can speed it up. Seems crazy to have to loop over each entry times by the number of series.

If 200 series and 100 entry rss feed that's 20,000 loops

**Comment by cvium** [comment]
Could we not sort the configured series alphabetically and put them in a dict keyed on first letter like this
```
series:
  - a series
  - another series
  - other series
  - third series
```
-->
```
series_map = {
    'a': ['a series', 'another series'],
    'o': ['other series'],
    't': ['third series']
}
```

**Comment by stevezau** [comment]
How will that help @cvium?

Won’t we still need to loop over them all right?

**Comment by cvium** [comment]
We can guess the series name from the title, grab the first letter and look up in the dict and try to match with the (hopefully much shorter) list

**Comment by stevezau** [comment]
Hmm is that safe to do? I guess we need to also check alt names..

**Comment by cvium** [comment]
All possible names would need to have its own node in the tree

**Comment by stevezau** [comment]
@gazpachoking i'm getting lost when trying to figure out how to properly impl series.name.in_ 

Do you understand how Comparator work? I think we need to impl def in_() but i'm not sure what to return?

**Comment by gazpachoking** [comment]
@stevezau Maybe try this for the comparator:
```python
def in_(other):
    return super().in_([normalize_series_name(e) for e in other])
```

**Comment by stevezau** [comment]
@gazpachoking that's what i thought yesterday but it still calls operate https://github.com/Flexget/Flexget/blob/series_speedup/flexget/plugins/filter/series.py#L238

I can fix it by checking if it's a list in operate but that does not seem right?

File: flexget/plugins/filter/series.py [hunk 1]
```diff
@@ -1579,13 +1592,36 @@ def auto_exact(self, config):
...
-            log.trace('series_name: `%s`, series_config: `%s`', series_name, series_config)
+
+        parser = get_plugin_by_name('parsing').instance
+
+        # Sort Entries into data model similar to https://en.wikipedia.org/wiki/Trie
+        # Only process series if both the entry title and series title first letter match
+        entries_map = {}
+        for entry in task.entries:
+            parsed = parser.parse_series(entry['title'])
+            if parsed.name:
```
Comment (liiight) [inline flexget/plugins/filter/series.py]: maybe also check `parsed.valid`, not sure if it'll have a `name` property if it isn't valid

File: flexget/plugins/filter/series.py [hunk 2]
```diff
@@ -1579,13 +1592,36 @@ def auto_exact(self, config):
...
         self.auto_exact(config)
-        for series_item in config:
-            series_name, series_config = list(series_item.items())[0]
-            log.trace('series_name: `%s`, series_config: `%s`', series_name, series_config)
+
+        parser = get_plugin_by_name('parsing').instance
+
+        # Sort Entries into data model similar to https://en.wikipedia.org/wiki/Trie
+        # Only process series if both the entry title and series title first letter match
+        entries_map = {}
```
Comment (liiight) [inline flexget/plugins/filter/series.py]: you can set it to `defaultdict(list)`, makes for a neater implementation later

File: flexget/plugins/filter/series.py [hunk 3]
```diff
@@ -1579,13 +1592,36 @@ def auto_exact(self, config):
...
+            parsed = parser.parse_series(entry['title'])
+            if parsed.name:
+                entries_map.setdefault(parsed.name[:1].lower(), []).append(entry)
+
+        with Session() as session:
+            # Preload series
+            # str() added to make sure number shows (e.g. 24) are turned into strings
+            series_names = [str(s.keys()[0]) for s in config]
+            existing_db_series = session.query(Series).filter(Series.name.in_(series_names))
+            existing_db_series = dict([(s.name_normalized, s) for s in existing_db_series])
```
Comment (liiight) [inline flexget/plugins/filter/series.py]: `existing_db_series = {s.name_normalized: s for s in existing_db_series}`

File: flexget/utils/tools.py [hunk 4]
```diff
@@ -482,6 +482,19 @@ def get_config_hash(config):
...
 
+def get_config_as_array(config, key):
+    """
+    Return configuration key as array, even if given as a single string
+    :param dict config: Configuration
+    :param string key: Configuration
+    :return: Array
+    """
+    v = config.get(key, [])
+    if isinstance(v, str):
```
Comment (liiight) [inline flexget/utils/tools.py]: this should probably be `if isinstance(v, basestring)`

**Comment by stevezau** [comment]
@gazpachoking @liiight @cvium tests are passing. Ready for review.

File: flexget/plugins/filter/series.py [hunk 5]
```diff
@@ -2158,16 +2218,27 @@ def on_task_start(self, task, config):
                     db_series.alternate_names = [alt for alt in db_series.alternate_names if alt.alt_name in alts]
                     # Add/update the possibly new alternate names
                 else:
+                    # TODO: Remove, added for debugging
```
Comment (cvium) [inline flexget/plugins/filter/series.py]: Forget something @stevezau ?

**Pull request merged**